File: kotlin-native/performance/KotlinVsSwift/ring/src/IntStreamBenchmark.swift
/*
 * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
 * that can be found in the LICENSE file.
 */

import Foundation

class IntStreamBenchmark {
    private var _data: AnySequence<Int>? = nil
    var data: AnySequence<Int> {
        return _data!
    }

    init() {
        _data = AnySequence(intValues(Constants.BENCHMARK_SIZE))
    }

    func copy() -> [Int] {
        return Array(data)
    }
    
    func copyManual() -> [Int] {
        var list: [Int] = []
        for item in data.lazy {
            list.append(item)
        }
        return list
    }
    
    func filterAndCount() -> Int {
        return data.lazy.filter { filterLoad($0) }.count
    }
    
    func filterAndMap() {
        for item in (data.lazy.filter { filterLoad($0) }.map { mapLoad($0) }) {
            Blackhole.consume(item)
        }
    }
    
    func filterAndMapManual() {
        for it in data.lazy {
            if (filterLoad(it)) {
                let item = mapLoad(it)
                Blackhole.consume(item)
            }
        }
    }
    
    func filter() {
        for item in (data.lazy.filter { filterLoad($0) }) {
            Blackhole.consume(item)
        }
    }
    
    func filterManual() {
        for it in data.lazy {
            if (filterLoad(it)) {
                Blackhole.consume(it)
            }
        }
    }
    
    func countFilteredManual() -> Int {
        var count = 0
        for it in data.lazy {
            if (filterLoad(it)) {
                count += 1
            }
        }
        return count
    }
    
    func countFiltered() -> Int {
        return data.lazy.count { filterLoad($0) }
    }
    
    func countFilteredLocal() -> Int {
        return data.lazy.cnt { filterLoad($0) }
    }
    
    func reduce() -> Int {
        return data.lazy.reduce(0) { if (filterLoad($1)) { return $0 + 1 } else {return $0 } }
    }
}

extension LazySequence {
    func count(where test: (Element) throws -> Bool) rethrows -> Int {
        return try self.filter(test).count
    }
}

File: kotlin-native/performance/ring/src/commonMain/kotlin/org/jetbrains/ring/IntStreamBenchmark.kt
/*
 * Copyright 2010-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.ring

import kotlinx.benchmark.*
import org.jetbrains.benchmarksLauncher.SkipWhenBaseOnly

private const val BENCHMARK_SIZE = 10000

@State(Scope.Benchmark)
@Measurement(time = 100, timeUnit = BenchmarkTimeUnit.MILLISECONDS)
class IntStream : SkipWhenBaseOnly() {
    private var _data: Iterable<Int>? = null
    val data: Iterable<Int>
        get() = _data!!

    init {
        _data = intValues(BENCHMARK_SIZE)
    }

    @Benchmark
    fun copy(bh: Blackhole) {
        skipWhenBaseOnly()
        bh.consume(data.asSequence().toList())
    }

    @Benchmark
    fun copyManual(bh: Blackhole) {
        skipWhenBaseOnly()
        val list = ArrayList<Int>()
        for (item in data.asSequence()) {
            list.add(item)
        }
        bh.consume(list)
    }

    @Benchmark
    fun filterAndCount(bh: Blackhole) {
        skipWhenBaseOnly()
        bh.consume(data.asSequence().filter { filterLoad(it) }.count())
    }

    @Benchmark
    fun filterAndMap(bh: Blackhole) {
        skipWhenBaseOnly()
        var result = 0
        for (item in data.asSequence().filter { filterLoad(it) }.map { mapLoad(it) })
            result += item.length
        bh.consume(result)
    }

    @Benchmark
    fun filterAndMapManual(bh: Blackhole) {
        skipWhenBaseOnly()
        var result = 0
        for (it in data.asSequence()) {
            if (filterLoad(it)) {
                result += mapLoad(it).length
            }
        }
        bh.consume(result)
    }

    @Benchmark
    fun filter(bh: Blackhole) {
        skipWhenBaseOnly()
        var result = 0
        for (item in data.asSequence().filter { filterLoad(it) })
            result += item
        bh.consume(result)
    }

    @Benchmark
    fun filterManual(bh: Blackhole){
        skipWhenBaseOnly()
        var result = 0
        for (it in data.asSequence()) {
            if (filterLoad(it))
                result += it
        }
        bh.consume(result)
    }

    @Benchmark
    fun countFilteredManual(bh: Blackhole) {
        skipWhenBaseOnly()
        var count = 0
        for (it in data.asSequence()) {
            if (filterLoad(it))
                count++
        }
        bh.consume(count)
    }

    @Benchmark
    fun countFiltered(bh: Blackhole) {
        skipWhenBaseOnly()
        bh.consume(data.asSequence().count { filterLoad(it) })
    }

    @Benchmark
    fun countFilteredLocal(bh: Blackhole) {
        skipWhenBaseOnly()
        bh.consume(data.asSequence().cnt { filterLoad(it) })
    }

    @Benchmark
    fun reduce(bh: Blackhole) {
        skipWhenBaseOnly()
        bh.consume(data.asSequence().fold(0) {acc, it -> if (filterLoad(it)) acc + 1 else acc })
    }
}
